core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -126,13 +125,48 @@ class KafkaApis(val requestChannel: RequestChannel,
     info("Shutdown complete.")
   }
 
+  private def validateForwardRequest(request: RequestChannel.Request): Unit = {
+    if (!config.forwardingEnabled) {
+      throw new InvalidRequestException(s"The handling of forwarding request $request is not enabled on this broker.")
+    } else if (!request.context.fromPrivilegedListener ||
+      !authorize(request.envelopeContext.get.brokerContext, CLUSTER_ACTION, CLUSTER, CLUSTER_NAME)) {
+      // If the designated forwarding request is not coming from a privileged listener, or
+      // it fails CLUSTER_ACTION permission, we would fail the authorization.
+      throw new ClusterAuthorizationException(s"Envelope request $request is not authorized")
+    } else if (!request.header.apiKey.forwardable) {
+      throw new InvalidRequestException("Given RPC " + request.header.apiKey + " does not support forwarding")
+    } else if (request.principalSerde.isEmpty) {
+      throw new PrincipalDeserializationFailureException("Principal serde is not defined for forwarding request")
+    } else if (!controller.isActive) {
+      throw new NotControllerException("Non-active controller could not process forward request")
+    }
+  }
+
+  private def maybeForward(request: RequestChannel.Request,
+                           handler: RequestChannel.Request => Unit): Unit = {
+    if (!request.isForwarded() && !controller.isActive && isForwardingEnabled(request)) {
```

Comment:
Perhaps it is obvious, but this logic does not give us any tight guarantees that the request will actually be handled by the broker that is currently the controller. For example, a new controller might get elected between the check in `validateForwardRequest` and the handler here. That is probably fine at the moment, because the zk logic in `AdminManager` can execute on any broker. 

If we imagine instead how this will work with the kip-500 controller, I think the incoming request will get put on the controller's queue. By the time the request gets dequeued, we will be able to know for sure whether this node is the controller or not, so we will be able to have a much better guarantee.

The only reason I bring this up is that we are currently assuming that the NOT_CONTROLLER gets propagated in the envelope error field. We'll have to keep this in mind when we adapt this logic for the new controller.